# BME-160
Research Programming taught by Professor David Bernick at UCSC. 

Lab 1: Introduction to Python

    "HelloMe": Prints out Hello along with the name of the user. 

    "HelloAgain": Prints a greeting given 2 parameters

    "myBio": Prints out the information of the user given certain parameters. 

    "countAT-bugs": Gives the AT content (calculated by counting the number of adenine and thymine within the DNA sequences and divided by the length of the DNA sequence) given a sequence of DNA nucleotides. 

Lab 2: Manipulating Data Types
     
     "seqCleaner": Removes ambiguous nucleotides within a sequence and replaces with a cont of how many ambiguous nucloetides were found within the sequence.  
     
     "fastqParse": Parses through a FASTQ file and outputs certain values which are: Instrument, Run ID, Flow Cell ID, Flow Cell Lane,Tile Number, X-Coord, Y-Coord
     
     "coordinatorMathSoln": Asks for and collects 3 sets of coordinates and outputs the N-C and N-Ca bond lengths and the C-N-CA bond angle (in degrees) with correct number of significant digits
     
     "converter": Convert a DNA sequence to the amino acid representation using a dictionary. 

Lab 3: Flow Control
    
    "proteinParams": Reads in a protein sequence and prints out the number of amino acids, total molecular weight, molar extinction coefficient, mass extinction coefficient, theoretical isoelectric point, and amino acid composition. 

Lab 4: Files and Modules

    "seqeunceAnalysis": Contained 4 different classes that looka at the amino acid composition, nucleotide composition, codon composition, sequence addition, and number of nucleotides. 
    
    "genomeAnalyzer": Determines which codons are preferred for each of the amino acids and calculating the relative percentage.
    

Lab 5: ORF Finding

    "findORFs": Analyzes a FASTA file in order to find ORFS, or start and end codons, that are longer than 100 nucleotides. 

Lab 6: Find Unique

    "findUnique": contains a python command line that reads a file of fasta sequences from STDIN, finds unqiue subsequences that occur in each tRNA. 
